File: src/binary-search/cpp/rbsearch.cpp
/**
 * Script para apresentar um exemplo de utilização da implementação recursiva da busca binária
*/

#include <vector>
#include <iostream>
#include "search.h"

/**
 * Função auxiliar para facilitar a exibição dos resultados
 */
void printSearchResult(int element, int result) {
    cout << "O elemento " << element << " está na posição: " << result << endl;
}

int main() {
    vector<int> elements = {
        55, 56, 57, 89, 101, 102, 103, 506, 708, 1001
    };

    int elementToSearch = 103;
    int elementToSearchTwo = 10000;

    printSearchResult(elementToSearch, recursiveBinarySearch(elements, 0, elements.size(), elementToSearch));
    printSearchResult(elementToSearchTwo, recursiveBinarySearch(elements, 0, elements.size(), elementToSearchTwo));
}

File: src/binary-search/lua/rbsearch.lua
--[[
    Exemplo de algoritmo de busca binária (Recursivo) implementado em Lua
]]

function binarySearchFunction( list, elementToSearch, startIndex, endIndex )
    halfIndex = (startIndex + endIndex) // 2

    if (startIndex > endIndex) then
        return -1
    end

    if (list[halfIndex] == elementToSearch) then
        return halfIndex
    end
    
    if (list[halfIndex] > elementToSearch) then
        -- O -1 é utilizado para 'remover' o meio que já foi verificado
        return binarySearchFunction(list, elementToSearch, startIndex, halfIndex - 1)
    else
        -- A mesma ideia utiliza acima tem aplicação aqui, com o diferencial de ser +1
        -- Já que a ideia é eliminar o lado já verificado
        return binarySearchFunction(list, elementToSearch, halfIndex + 1, endIndex)
    end
end

function recursiveBinarySearch( list, elementToSearch )
    startIndex = 1
    endIndex = #list

    return binarySearchFunction(list, elementToSearch, startIndex, endIndex)
end

elements = {
    5, 7, 8, 9, 11, 12, 15, 77, 88, 99, 101
}

print('Buscando elemento existente na lista')
print(recursiveBinarySearch(elements, 88))

print('Buscando elemento que não existe na lista')
print(recursiveBinarySearch(elements, 999))
